File: packages/gbnf-experiment/data/20260909T182714Z_2643e4be/ported_implementation/src/grammar-graph/get-parent-stack-id.ts
import { Color } from "./colorize.js";
import type { GraphPointer } from "./graph-pointer.js";

export const getParentStackId = (
  pointer: GraphPointer,
  col: (text: string | number, color: string) => string,
): string => {
  const stackIds: string[] = [];
  let parent: GraphPointer | null = pointer.parent;
  while (parent) {
    stackIds.push(
      `${parent.node.meta.stackId},${parent.node.meta.pathId},${parent.node.meta.stepId}`,
    );
    parent = parent.parent;
  }
  const arrow = col("<-", Color.GRAY);
  return stackIds.map((stackId) => col(stackId, Color.RED)).join(arrow);
};

File: proposed-projects/round-trip-experiment/reverse/20260909T130706Z_31ac2789/20260910T073404Z_10c90cd6/ported_implementation/gbnf/grammar_graph/get_parent_stack_id.py
from typing import Callable, List

from .colorize import Color


def get_parent_stack_id(pointer, col: Callable[..., str]) -> str:
    stack_ids: List[str] = []
    parent = pointer.parent
    while parent:
        stack_ids.append(
            f"{parent.node.meta['stackId']},"
            f"{parent.node.meta['pathId']},"
            f"{parent.node.meta['stepId']}"
        )
        parent = parent.parent
    arrow = col("<-", Color.GRAY)
    return arrow.join(col(stack_id, Color.RED) for stack_id in stack_ids)


getParentStackId = get_parent_stack_id
